### PR TITLE
Add uninstall command

### PR DESCRIPTION
**Pull request by nolancon**
Uninstall command (`kubectl storageos uninstall`) added:

`pkg/install` changed to `pkg/installer` and is responsible for both install and uninstall functionality.

More detail in commit comments.

### CODE REVIEW

File: pkg/utils/k8s.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+	if err != nil {
+		return err
+	}
+	podClient := clientset.CoreV1().Pods(namespace)
+	pods, err := podClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if len(pods.Items) > 0 {
+		return fmt.Errorf("pods still exist in namespace %s", namespace)
```
Comment (mhmxs) [inline pkg/utils/k8s.go:186]: I understand the reason of the single return value, but for me a bit annoying. The function knows about the outer world and can made the decision is it an error or not.

File: pkg/utils/yaml.go [hunk 2]
```diff
@@ -1,142 +1,101 @@
...
-	return config, nil
-}
-
-// GetClientsetFromConfig returns a k8s clientset from a client-go rest config
-func GetClientsetFromConfig(config *rest.Config) (*kubernetes.Clientset, error) {
-	clientset, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return nil, err
+	for _, obj := range objs.Items {
+		if obj.UnstructuredObject().GetKind() == kind {
```
Comment (mhmxs) [inline pkg/utils/yaml.go:23]: `if obj.UnstructuredObject().GetKind() != kind { continue }` should be a bit more goish.

File: pkg/installer/installer.go [hunk 3]
```diff
@@ -500,3 +356,54 @@ func (in *Installer) getManifestFromFsMultiDoc(path, kind string) (string, error
...
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// omitAndReturnKindFromFSMultiDoc reads the file at path of the in-memory filesystem, uses
+// OmitAndReturnKindFromMultiDoc internally to perform the update and then writes the returned file to path,
+// also returninng a []string of the objects omitted.
+func (in *Installer) omitAndReturnKindFromFSMultiDoc(path, kind string) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (nolancon) [reply pkg/installer/installer.go:417]: In fact the function it is duplicating is no longer used so just removing that one instead

File: pkg/installer/uninstall.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+	err := in.uninstallStorageOS(v.GetString(StosOperatorNSFlag), v.GetString(StosClusterNSFlag))
+	if err != nil {
+		return err
+	}
+
+	// return early if user only wishes to delete storageos, leaving etcd untouched
+	if v.GetBool(SkipEtcdFlag) {
+		return nil
+	}
+	err = in.uninstallEtcd(v.GetString(EtcdNamespaceFlag))
```
[the remark this answers is not in the document]
Comment (nolancon) [reply pkg/installer/uninstall.go:25]: I had been doing everything sequentially as it was easier to see what was happening while still in development stage. But maybe you are right - I will add this to my TODO list if that is OK

File: pkg/installer/uninstall.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	return nil
+}
+
+func (in *Installer) uninstallStorageOS(stosOperatorNS, stosClusterNS string) error {
+	var err error
+	// add changes to storageos kustomizations here before kustomizeAndDelete calls ie make changes
+	// to storageos/operator/kustomization.yaml and/or storageos/cluster/kustomization.yaml
+	// based on flags (or cli config file)
+	if stosOperatorNS != "" {
+		err := in.setFieldInFsManifest(filepath.Join(stosDir, operatorDir, kustomizationFile), stosOperatorNS, "namespace", "")
```
[the remark this answers is not in the document]
Comment (nolancon) [reply pkg/installer/uninstall.go:39]: I think the tight coupling you are referring to is in fact with the plugin's "config file"  (maybe you mean the same thing?).
Anyway, in majority of cases there will be no namespace specified in the _config_ and so no transform is required - therefore we fall back to the _manifest_ namespace. Also this code is changed around a bit in the Upgrade PR so I am inclined to leave as is for now.

File: pkg/installer/uninstall.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	err = in.kubectlClient.Delete(context.TODO(), "", string(manifest), true)
+	if err != nil {
+		return err
+	}
+
+	// gracefully delete removed namespaces (there is likely only one)
+	for _, removedNamespace := range removedNamespaces {
+		err = in.gracefullyDeleteNS(removedNamespace)
```
[the remark this answers is not in the document]
Comment (nolancon) [reply pkg/installer/uninstall.go:152]: Again, I might add this to my TODO list 😄

File: pkg/utils/k8s.go [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+	storageV1Client, err := storagev1.NewForConfig(restConfig)
+	if err != nil {
+		return "", err
+	}
+	storageClasses, err := storageV1Client.StorageClasses().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+	for _, storageClass := range storageClasses.Items {
+		for k, v := range storageClass.GetObjectMeta().GetAnnotations() {
```
[the remark this answers is not in the document]
Comment (nolancon) [reply pkg/utils/k8s.go:107]: 🔥 Done

File: pkg/utils/k8s.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+			if err == nil {
+				return nil
+			}
+		}
+	}
+}
+
+// DeploymentIsReady attempts to `get` a deployment by name and namespace, the function returns no error
+// if no deployment replicas are ready.
+func DeploymentIsReady(config *rest.Config, name, namespace string) error {
```
[the remark this answers is not in the document]
Comment (nolancon) [reply pkg/utils/k8s.go:139]: Done

File: pkg/utils/k8s.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+	}
+	if dep.Status.ReadyReplicas == 0 {
+		return fmt.Errorf("no replicas are ready for deployment %s; %s", name, namespace)
+	}
+	return nil
+}
+
+// PodIsRunning attempts to `get` a pod by name and namespace, the function returns no error
+// if the pod is in running phase.
+func PodIsRunning(config *rest.Config, name, namespace string) error {
```
[the remark this answers is not in the document]
Comment (nolancon) [reply pkg/utils/k8s.go:158]: Done

File: pkg/utils/k8s.go (hunk 1 above)
Comment (nolancon) [reply pkg/utils/k8s.go:186]: I think I need to understand better why this is annoying for you 😄 Maybe we can revisit this also and you can show me a nicer way of doing it.

File: pkg/utils/yaml.go (hunk 2 above)
Comment (nolancon) [reply pkg/utils/yaml.go:23]: Done

**Comment by nolancon** [comment]
Thanks @mhmxs for your review. I've added the simpler changes. The others, I will come back to after the upgrade command is merged and the code is stabilized a little bit if that is OK. 
Are you OK to merge this PR as is?
There are more changes to come in #13 so we can address any issues there, for now I'd like to get this merged so there is no big rebase as #13 is dependent on the commits in this PR 😄

**Comment by nolancon** [comment]
> @nolancon Yes I'm ok. Thanks

@mhmxs Great, please go ahead then. I am powerless without your approval 🧙‍♂️ 😆

**Pull request merged**